Framework/DataHandling/src/MaskDetectors.cpp

```diff
@@ -454,8 +472,38 @@ void MaskDetectors::appendToIndexListFromWS(
     }
   }
   tmp_index.swap(indexList);
+}
 
-} // appendToIndexListFromWS
+/**
+* Append the indices of a workspace corresponding to detector IDs to the
+* given list
+*
+* @param detectorList :: An existing list of detector IDs
+* @param inputWs :: A workspace to mask detectors in
+* @param maskWs :: A workspace with masked spectra
+* @param range_info    :: tuple containing the information on
+*                      if copied indexes are constrained by ranges and if yes
+*                       -- the range of indexes to topy
+*/
+void MaskDetectors::appendToDetectorListFromWS(
+    std::vector<detid_t> &detectorList,
+    const MatrixWorkspace_const_sptr inputWs,
+    const MatrixWorkspace_const_sptr maskWs,
+    const std::tuple<size_t, size_t, bool> &range_info) {
+  size_t startIndex = std::get<0>(range_info);
+  size_t endIndex = std::get<1>(range_info);
+
+  const auto &detMap = inputWs->getDetectorIDToWorkspaceIndexMap();
+  detectorList.reserve(maskWs->getNumberHistograms());
+
+  for (size_t i = 0; i < maskWs->getNumberHistograms(); ++i) {
+    if (maskWs->y(i)[0] == 0) {
+      auto id = maskWs->getDetector(i)->getID();
```

Comment:
There is a methods that will return a vector of all of the IDs (i think getIDs) that will allow this to work for none 1-1 mapping.